Serialization of <position> argument of CSS circle() and ellipse() should match spec

weinig opened the pull request:
#### f456176de25fc0c1be376d68c40732360d8030ce
<pre>
Serialization of &lt;position&gt; argument of CSS circle() and ellipse() should match spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=279200">https://bugs.webkit.org/show_bug.cgi?id=279200</a>

Reviewed by Darin Adler.

The current serialization of the &lt;position&gt; argument to CSS circle() and ellipse()
was not following the spec, which outlines exactly how specified and computed
serialization should behave.

This updates the implementation and quite a few tests to align.

* LayoutTests/fast/masking/parsing-clip-path-shape-expected.txt:
* LayoutTests/fast/masking/parsing-clip-path-shape.html:
* LayoutTests/fast/shapes/parsing/parsing-shape-lengths-expected.txt:
* LayoutTests/fast/shapes/parsing/parsing-shape-lengths.html:
* LayoutTests/fast/shapes/parsing/parsing-shape-outside-expected.txt:
* LayoutTests/fast/shapes/parsing/parsing-test-utils.js:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/basic-shape-circle-ellipse-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/basic-shape-circle-ellipse-serialization.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-valid-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-valid-position.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/circle-function-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/ellipse-function-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-009.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-011.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-009.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/support/parsing-utils.js:
    - Update tests to test &lt;position&gt; serialization rules.

* Source/WebCore/rendering/style/BasicShapes.cpp:
* Source/WebCore/rendering/style/BasicShapes.h:
* Source/WebCore/css/BasicShapeConversion.cpp:
(WebCore::valueForCenterCoordinate):
(WebCore::valueForBasicShape):
(WebCore::convertToCenterCoordinate):
(WebCore::floatValueForCenterCoordinate):
    - Simplify BasicShapeCenterCoordinate to just store a Length. This can always be fully computed
      at style building time and is not expected to retain keywords for computed value serialization.

* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::CSSCircleValue::customCSSText const):
(WebCore::CSSEllipseValue::customCSSText const):
(WebCore::serializePositionOffset): Deleted.
(WebCore::isZeroLength): Deleted.
(WebCore::buildSerializablePositionOffset): Deleted.
    - Serialization of &lt;position&gt; is now just a direct serialization of the x/y arguments.
      At parse time, we already ensure hat if only one is defined, the other gets set to
      CSSValueCenter. And at compute value time, we ensure they are always set to the computed
      Length values.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update serialization for simplified BasicShapeCenterCoordinate.

Canonical link: <a href="https://commits.webkit.org/283287@main">https://commits.webkit.org/283287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fbba98779d4d9c060eb6207063ad377a5d2218e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65856 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18475 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/69885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67974 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16747 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/69885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/11438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68923 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/57006 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/69885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/38421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15341 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71588 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9811 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9843 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/57073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14511 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8104 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43296 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41857 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/BasicShapeConversion.cpp:
```diff
@@ -48,14 +48,9 @@
 namespace WebCore {
 
-static Ref<CSSValue> valueForCenterCoordinate(const RenderStyle& style, const BasicShapeCenterCoordinate& center, BoxOrient orientation)
+static Ref<CSSValue> valueForCenterCoordinate(const RenderStyle& style, const BasicShapeCenterCoordinate& center)
```
Do we still need a function for this?

weinig replied:
Nope. Gone.

darinadler commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -88,45 +88,26 @@ class BasicShape : public RefCounted<BasicShape> {
-    Direction m_direction { Direction::TopLeft };
-    Length m_length { LengthType::Undefined };
-    Length m_computedLength;
+    Length m_length { Length(100, LengthType::Percent) };
```
Do we need `Length()`?

weinig committed:
Serialization of <position> argument of CSS circle() and ellipse() should match spec
https://bugs.webkit.org/show_bug.cgi?id=279200

Reviewed by Darin Adler.

The current serialization of the <position> argument to CSS circle() and ellipse()
was not following the spec, which outlines exactly how specified and computed
serialization should behave.

This updates the implementation and quite a few tests to align.

* LayoutTests/fast/masking/parsing-clip-path-shape-expected.txt:
* LayoutTests/fast/masking/parsing-clip-path-shape.html:
* LayoutTests/fast/shapes/parsing/parsing-shape-lengths-expected.txt:
* LayoutTests/fast/shapes/parsing/parsing-shape-lengths.html:
* LayoutTests/fast/shapes/parsing/parsing-shape-outside-expected.txt:
* LayoutTests/fast/shapes/parsing/parsing-test-utils.js:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/basic-shape-circle-ellipse-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/basic-shape-circle-ellipse-serialization.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-valid-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/parsing/shape-outside-valid-position.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/circle-function-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/ellipse-function-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-009.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-circle-011.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-ellipse-009.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/support/parsing-utils.js:
    - Update tests to test <position> serialization rules.

* Source/WebCore/rendering/style/BasicShapes.cpp:
* Source/WebCore/rendering/style/BasicShapes.h:
* Source/WebCore/css/BasicShapeConversion.cpp:
(WebCore::valueForCenterCoordinate):
(WebCore::valueForBasicShape):
(WebCore::convertToCenterCoordinate):
(WebCore::floatValueForCenterCoordinate):
    - Simplify BasicShapeCenterCoordinate to just store a Length. This can always be fully computed
      at style building time and is not expected to retain keywords for computed value serialization.

* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::CSSCircleValue::customCSSText const):
(WebCore::CSSEllipseValue::customCSSText const):
(WebCore::serializePositionOffset): Deleted.
(WebCore::isZeroLength): Deleted.
(WebCore::buildSerializablePositionOffset): Deleted.
    - Serialization of <position> is now just a direct serialization of the x/y arguments.
      At parse time, we already ensure hat if only one is defined, the other gets set to
      CSSValueCenter. And at compute value time, we ensure they are always set to the computed
      Length values.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update serialization for simplified BasicShapeCenterCoordinate.

Canonical link: https://commits.webkit.org/283287@main
